### PR TITLE
Added functionality for multiple base domains

### PR DESCRIPTION
**Pull request by FrankGasparovic**
US49187:
- Added functionality for multiple base domains
- Added test for multiple base domains

### CODE REVIEW

**Comment by FrankGasparovic** [comment]
[Build Status](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/token-lib-dev-pullrequest/102/)

Build Success

File: src/main/java/com/ge/predix/uaa/token/lib/AbstractZoneAwareTokenService.java [hunk 1]
```diff
@@ -212,13 +211,17 @@ public void setServiceZoneHeaders(final String serviceZoneHeaders) {
         this.serviceZoneHeaders = serviceZoneHeaders;
     }
 
-    /* Package Private */ List<String> getServiceZoneHeaderList() {
-        return Arrays.asList(this.serviceZoneHeaders.split(","));
+            /* Package Private */ List<String> splitCSV(final String csvString) {
```
Comment (sanjeevchopra) [inline src/main/java/com/ge/predix/uaa/token/lib/AbstractZoneAwareTokenService.java]: this method should be private & pls fix indent

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java [hunk 2]
```diff
@@ -107,6 +107,15 @@ public void testGetSubdomainFromOneHeader(final String requestHeader, final List
         Assert.assertEquals(subdomain, subdomainValue);
     }
 
+    @Test(dataProvider = "domainDataProvider")
+    public void testGetSubdomainFromDomain(final String serverName, final List<String> serviceConfigBaseDomains,
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: please name this testGetZoneName , since the method it tests is getZoneName

Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: name suggestion for first parameter:   requestHostname
2nd param:   serviceBaseDomains

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java [hunk 3]
```diff
@@ -121,4 +130,12 @@ public void testGetSubdomainFromOneHeader(final String requestHeader, final List
                 { "unrecognized-headaer", Arrays.asList("Predix-Zone-Id", "ACS-Zone-Subdomain"), "", null },
                 { "Predix-Zone-Id", Arrays.asList("Unrecognized-Header"), "some-value", null } };
     }
+
+    @DataProvider(name = "domainDataProvider")
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: 👍 
minor:  use Collections.emptyList() to avoid creating new objects.

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java (hunk 2 above)
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: please add a new test to verify if 2 different zoneIds are specified in subdomain and header, then the one in subdomain is returned.

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 4]
```diff
@@ -215,11 +215,11 @@ private void assertAuthentication(final OAuth2Authentication authentication, fin
     public void testGetServiceHeaders() {
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: pls add a new test which tries to use ZacTokenService with neither headers or subdomain defined, it should throw IllegalStateException or similar

**Comment by FrankGasparovic** [comment]
Fixed 
Build Success
[Build Status](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/token-lib-dev-pullrequest/103/)

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 5]
```diff
@@ -131,16 +142,18 @@ public void testDefaultAndZoneSpecificResourceAuthorization(final String zoneId,
 
     }
 
-    private OAuth2Authentication loadAuthentication(final String zoneName, final String zoneUserScope) {
-        return loadAuthentication(zoneName, zoneUserScope, "/test/resource", Arrays.asList("/zone/**"));
+    private OAuth2Authentication loadAuthentication(String headers, String baseDomains, final String zoneName,
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: pleaes rename params:    configuredHeaderNames, configuredBaseDomains, requestZoneName, userScopes

Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: Consider splitting this method into smaller methods:    mockFastTokenService  , configureZacTokenService, mockHttpRequest ?

**Comment by FrankGasparovic** [comment]
Build Success

[Build Status](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/token-lib-dev-pullrequest/)

**Comment by FrankGasparovic** [comment]
Test successful

[Build Status](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/token-lib-dev-pullrequest/105/)

File: src/main/java/com/ge/predix/uaa/token/lib/AbstractZoneAwareTokenService.java [hunk 6]
```diff
@@ -66,9 +67,13 @@
     public OAuth2Authentication loadAuthentication(final String accessToken)
             throws AuthenticationException, InvalidTokenException {
 
+        if (this.getServiceBaseDomain().isEmpty() && this.getServiceZoneHeadersList().isEmpty()) {
+            throw new InvalidRequestException(
+                    "Service configuration error: Service not providing base domain or header parameter");
```
Comment (sanjeevchopra) [inline src/main/java/com/ge/predix/uaa/token/lib/AbstractZoneAwareTokenService.java]: "ZacTokenService requires atleast one of the following properties to be configured: serviceBaseDomain or serviceZoneHeaders ."

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java [hunk 7]
```diff
@@ -98,27 +99,50 @@ public void getSubdomainEmptyHostname() {
...
+                // domains only configured
+                { "zone1.acs.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com"), Collections.emptyList(),
+                        "zone1" },
+                { "zone1.acs.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com", "guardians.com"),
+                        Collections.emptyList(), "zone1" },
+                { "zone2.guardians.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com", "guardians.com"),
+                        Collections.emptyList(), "zone2" },
+                { "guardians.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com", "guardians.com"),
+                        Collections.emptyList(), null },
+                { "guardians.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com", "guardians.com"),
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: duplicate of previous one ?

**Comment by FrankGasparovic** [comment]
[Build Status](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/token-lib-dev-pullrequest/107/) 

Jenkins Status Build Success

File: src/main/java/com/ge/predix/uaa/token/lib/ZacTokenService.java [hunk 8]
```diff
@@ -72,5 +93,10 @@ public void flushIssuerCache(final String zoneId) {
     public void afterPropertiesSet() throws Exception {
         long timeToLiveMillis = this.issuersTtlSeconds * 1000;
         this.tokenServicesMap = new PassiveExpiringMap<>(timeToLiveMillis);
+
```
Comment (sanjeevchopra) [inline src/main/java/com/ge/predix/uaa/token/lib/ZacTokenService.java]: this should suffice 
`if (  StringUtils.isBlank(this.serviceBaseDomain) || StringUtils.isBlank(this.serviceZoneHeaders) {
.....
} `

consider IllegalArgumentException, because in this case there is no request yet.

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 9]
```diff
@@ -182,6 +182,8 @@ private OAuth2Authentication loadAuthentication(String configuredHeaderNames, St
         zacTokenServices.setOauth2RestTemplate(restTemplateMock);
         try {
             zacTokenServices.afterPropertiesSet();
+        } catch (InvalidRequestException r) {
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: why is this needed ?

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java [hunk 10]
```diff
@@ -134,8 +134,6 @@ public void testGetZoneNameWithHeaderAndBaseDomains(final String requestHostname
                         Collections.emptyList(), "zone2" },
                 { "guardians.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com", "guardians.com"),
                         Collections.emptyList(), null },
-                { "guardians.com", "Predix-Zone-Id", "DONTUSE", Arrays.asList("acs.com", "guardians.com"),
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: still need a test for this case, which asserts that a exception is thrown
